docs/rfcs/033-pageserver-postgres-bundle.md

```diff
@@ -0,0 +1,139 @@
+3. some dependencies can cause database corruption if updated carelessly (locale => libc)
+
+## Why Is This Worth Solving
+
+1. Clearly defined ownership generally boosts execution speed & bug triage.
+   * Example for why execution speed matters: CVE in dependency => who takes care of patching & updating.
+2. Centralize understanding of risks involved with some dependencies.
+   Currently, there is no team clearly responsible for assessing / tracking the risks. As a reminder from previous section, these are
+   * runtime incompatibilities
+   * database corruption
+
+Also, it is an unlock for additional future value, see "Future Work" section.
+
+## Impacted components (e.g. pageserver, safekeeper, console, etc)
+
+Pageserver (neon.git)
+Compute (neon.git)
+Deployment process (aws.git)
+
+## Design
+
+The basic interface between Compute and Storage team is as follows:
+
+* Compute team publishes a "bundle" of the binaries required by Pageserver
+* Storage team uses a pinned bundle in the Pageserver build process
+* Storage team code review is required to update the pinned version
+
+The "bundle" provides an interface agreed upon by Compute and Storage teams to run
+* for each supported Postgres version at Neon (v14, v15, v16, ...)
+  * the `initdb` process
+    * behaving like a vanilla Postgres `initdb`
+  * `postgres --wal-redo` mode process
+    * following the walredo protocol specified elsewhere
+
+The bundle is self-contained, i.e., it behaves the same way on any Linux system.
+The only ambient runtime dependency is the Linux kernel.
+The minimum Linux kernel version is 5.10.
+
+### Variant 1: bundle = fully statically linked binaries
+The "bundle" is a tarball of fully statically linked binaries
+
+```
+v14/initdb
+v14/postgres
+v15/initdb
+v15/postgres
+v16/initdb
+v16/postgres
+...
+```
+
+The directory structure is part of the interface.
+
+### Variant 2: bundle = chrooted directory
+
+The "bundle" is a tarball that contains all sorts of files, plus a launcher script.
+
+```
+LAUNCHER
+storage
+storage/does
+storage/does/not
+storage/does/not/care
+```
+
+To launch `initdb` or `postgres --wal-redo`, the Pageserver does
+1. fork child process
+2. `chroot` into the extracted directory
+3. inside the chroot, run `/LAUNCHER VERSION PG_BINARY [FLAGS...]`
+4. The `LAUNCHER` script sets up library search paths, etc, and then `exec`s the correct binary
+
+We acknowledge this is half-way reinventing OCI + linux containers.
+However, our needs are much simpler than what OCI & Docker provide.
+Specifically, we do not want Pageserver to be runtime-dependent on e.g. Docker as the launcher.
+
+The `chroot` is to enforce that the "bundle" be self-contained.
+The special path `/inout` int he bundle is reserved, e.g., for `initdb` output.
+
+### Variant 3: ???
+
+Your design here, feedback welcome.
```

Comment:
I think pinning "wal-redo bundle" version has downsides -- now instead of building everything from main you would have to take some historic version of wal-redo. That is either juggling with git history during build to pick old version (would break builds with shallow repo) or taking some ready binaries that were built previously (we depend on some scripts/actions that are outside of repo). I'd keep building "wal-redo" the way we build it now -- each time pageserver is built.

IMO main benefit and risk reduction would be from more hermetic wal-redo binaries and avoiding initdb in PS.